### PR TITLE
docker image validation

### PR DESCRIPTION
**Pull request by danielnorberg**

### CODE REVIEW

**Comment by danielnorberg** [comment]
@rouzwawi PTAL

File: styx-api-service/src/main/java/com/spotify/styx/StyxApi.java [hunk 1]
```diff
@@ -109,7 +110,7 @@ public void create(Environment environment) {
     final Storage storage = storageFactory.apply(environment);
 
     final WorkflowResource workflowResource =
-        new WorkflowResource(storage, schedulerServiceBaseUrl);
+        new WorkflowResource(storage, schedulerServiceBaseUrl, new DockerImageValidator());
```
[the remark this answers is not in the document]
Comment (danielnorberg) [reply styx-api-service/src/main/java/com/spotify/styx/StyxApi.java]: Having `DockerImageValidator` be a regular class allows for mocking when testing.

File: styx-common/src/main/java/com/spotify/styx/util/DockerImageValidator.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+  private static final Pattern DIGIT_PERIOD = Pattern.compile("^[0-9.]+$");
+
+  public Collection<String> validateImageReference(final String imageRef) {
+    final Collection<String> errors = new HashSet<>();
+    validateImageReference(imageRef, errors);
+    return errors;
+  }
+
+  @SuppressWarnings("ConstantConditions")
+  private boolean validateImageReference(final String imageRef, final Collection<String> errors) {
```
[the remark this answers is not in the document]
Comment (danielnorberg) [reply styx-common/src/main/java/com/spotify/styx/util/DockerImageValidator.java]: Final all the things! \o/

[the remark this answers is not in the document]
Comment (honnix) [reply styx-common/src/main/java/com/spotify/styx/util/DockerImageValidator.java]: Java10 or 11 or whatever should invert the semantics.

[the remark this answers is not in the document]
Comment (danielnorberg) [reply styx-common/src/main/java/com/spotify/styx/util/DockerImageValidator.java]: Kotlin rewrite? ;)

**Comment by bergman** [comment]
👍

**Pull request merged**